c45.py

```diff
@@ -0,0 +1,217 @@
+        if diff_entropy < least_gain:
+            tree.right_branch, tree.left_branch = None, None
+            tree.results = n_distinct_dict(left + right)
+
+"""
+Helper functions: type_conversion, fread
+"""
+def type_conversion(val):
+        val = val.strip()
+        
+        try:
+            if '.' in val:
+                return float(val)
+            else:
+                return int(val)
+            
+        except ValueError:
+            #For other types, return
+            return val
+        
+def fread(f, col_labels = False):
+    """
+    takes a filepath, f, and a boolean argument, col_labels.
+    By default, col_labels is False, implying that the columns do not have labels. If set to true,
+    fread will remove the row containing column labels at the end.
+    """
+    data = csv.reader(open(f, 'rt'))
+    lst = [[type_conversion(i) for i in r] for r in data]
+    if col_labels:
+        lst.pop(0)
+    return lst
+
+
+def partition(r, c, val):
+    """
+    Function to partition the data based on value
+    """
+    #Declare anonymous function
+    split_fun = None
+    if isinstance(val, float) or isinstance(val, int): 
+        #Anonymous function for numeric values
+        split_fun = lambda row : row[c] >= val
+    else: 
+        #For string values
+        split_fun = lambda row : row[c] == val
+    list1 = [row for row in r if split_fun(row)]
+    list2 = [row for row in r if not split_fun(row)]
+    return (list1, list2)
+
+
+
+def train(lst, criteria = entropy):
+    """
+    Decision tree construction - by default, the entropy function is used to calculate the criteria for splitting. 
+    lst : dataframe with the last column reserved for labels
+    criteria : entropy or gini calculation function
+    """
+    #Base Case: Empty Set
+    if len(lst) == 0: 
+        return Tree()
+    
+    #Calculate Entropy/Gini of current X, declare A_best, create sets/gain accordingly
+    score = criteria(lst)
+    Attribute_best = None
+    Set_best = None
+    Gain_best = 0.0
+    
+
+    num_col = len(lst[0]) - 1  # last column of lst is labels
+    for c in range(num_col):
+        col_val = [row[c] for row in lst]
+        for value in col_val:
+            #Split dataset
+            (set1, set2) = partition(lst, c, value)
+            # Calculate Gain
+            p = float(len(set1))/len(lst)
+            gain = score - p*criteria(set1) - (1-p)*criteria(set2)
+            if gain>Gain_best and len(set1)>0 and len(set2)>0:
+                Gain_best = gain
+                Attribute_best = (c, value)
+                Set_best = (set1, set2)
+
+    if Gain_best > 0:
+        #Recursive Call on partitioned Sets
+        right_branch = train(Set_best[0])
+        left_branch = train(Set_best[1])
+        return Tree(col=Attribute_best[0], value=Attribute_best[1], right_branch=right_branch, left_branch=left_branch)
+    
+    else:
+        return Tree(results=n_distinct_dict(lst))
+
+
+def tree_classify(X, tree):
```

Comment:
the functions like predict, train, prune, classify, might be good for putting as methods for a decision tree model class; similar to how scikit does it, you would make a new instance of that model class, call a fit method (which trains and prunes to whatever specified height/bin size), and then call a predict method to make prediction. 

extra layer of abstraction would avoid exposing the structure of the tree and unnecessarily passing the tree object between diff functions (tree structure should still be saved in a field somewhere that way we can display it)